# Agent Analysis Notebook

**Task:** Retrieve financial data from the mcp.financialdatasets.ai API and perform a comparative analysis of major Big Tech AI companies over the past year. Focus on metrics such as revenue growth, R&D or AI-related investments, profitability, and market performance. Summarize key insights and generate visualizations to support the comparison.

**Generated:** 2026-01-02 14:47:22

**Agent Type:** Planner Agent (with dynamic task planning)

*This notebook was automatically cleaned: imports consolidated, failed cells removed.*

---



## Setup & Imports


In [1]:
from datetime import datetime
import warnings

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns


## Analysis


### Load and explore all retrieved data


In [2]:
warnings.filterwarnings('ignore')

# Set style for better-looking plots
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (14, 8)
plt.rcParams['font.size'] = 10

# Define the companies
companies = ['MSFT', 'AAPL', 'GOOGL', 'AMZN', 'META', 'NVDA']
company_names = {
    'MSFT': 'Microsoft',
    'AAPL': 'Apple',
    'GOOGL': 'Alphabet (Google)',
    'AMZN': 'Amazon',
    'META': 'Meta (Facebook)',
    'NVDA': 'NVIDIA'
}

# Create a summary dataframe from financial metrics
metrics_data = {
    'MSFT': {
        'Market Cap (B)': 3514.10,
        'P/E Ratio': 33.50,
        'Revenue Growth': 4.29,
        'Earnings Growth': 3.02,
        'Net Margin': 35.71,
        'Operating Margin': 44.30,
        'Gross Margin': 68.76,
        'ROE': 28.90,
        'ROA': 16.49,
        'EPS': 14.11,
        'Free Cash Flow Yield': 2.22
    },
    'AAPL': {
        'Market Cap (B)': 4016.12,
        'P/E Ratio': 35.86,
        'Revenue Growth': 1.84,
        'Earnings Growth': 12.82,
        'Net Margin': 26.92,
        'Operating Margin': 31.89,
        'Gross Margin': 46.91,
        'ROE': 151.91,
        'ROA': 31.18,
        'EPS': 7.55,
        'Free Cash Flow Yield': 2.46
    },
    'GOOGL': {
        'Market Cap (B)': 3816.54,
        'P/E Ratio': 30.72,
        'Revenue Growth': 3.79,
        'Earnings Growth': 7.51,
        'Net Margin': 32.23,
        'Operating Margin': 39.55,
        'Gross Margin': 59.17,
        'ROE': 32.12,
        'ROA': 23.16,
        'EPS': 10.27,
        'Free Cash Flow Yield': 1.93
    },
    'AMZN': {
        'Market Cap (B)': 2416.81,
        'P/E Ratio': 31.60,
        'Revenue Growth': 3.18,
        'Earnings Growth': 8.30,
        'Net Margin': 11.06,
        'Operating Margin': 13.76,
        'Gross Margin': 50.05,
        'ROE': 20.69,
        'ROA': 10.51,
        'EPS': 7.17,
        'Free Cash Flow Yield': 0.61
    },
    'META': {
        'Market Cap (B)': 1640.77,
        'P/E Ratio': 28.03,
        'Revenue Growth': 5.96,
        'Earnings Growth': -18.15,
        'Net Margin': 30.89,
        'Operating Margin': 44.41,
        'Gross Margin': 82.00,
        'ROE': 30.16,
        'ROA': 19.26,
        'EPS': 23.30,
        'Free Cash Flow Yield': 2.73
    },
    'NVDA': {
        'Market Cap (B)': 4603.51,
        'P/E Ratio': 46.41,
        'Revenue Growth': 13.27,
        'Earnings Growth': 14.55,
        'Net Margin': 53.01,
        'Operating Margin': 62.26,
        'Gross Margin': 70.05,
        'ROE': 83.43,
        'ROA': 61.56,
        'EPS': 4.08,
        'Free Cash Flow Yield': 1.68
    }
}

df_metrics = pd.DataFrame(metrics_data).T
print("=" * 80)
print("BIG TECH AI COMPANIES - KEY FINANCIAL METRICS SNAPSHOT")
print("=" * 80)
print(df_metrics.round(2))
print("\n")

# Save to CSV
df_metrics.to_csv('artifacts/financial_metrics_summary.csv')
print("✓ Saved financial metrics summary to artifacts/financial_metrics_summary.csv")


BIG TECH AI COMPANIES - KEY FINANCIAL METRICS SNAPSHOT
       Market Cap (B)  P/E Ratio  Revenue Growth  Earnings Growth  Net Margin  \
MSFT          3514.10      33.50            4.29             3.02       35.71   
AAPL          4016.12      35.86            1.84            12.82       26.92   
GOOGL         3816.54      30.72            3.79             7.51       32.23   
AMZN          2416.81      31.60            3.18             8.30       11.06   
META          1640.77      28.03            5.96           -18.15       30.89   
NVDA          4603.51      46.41           13.27            14.55       53.01   

       Operating Margin  Gross Margin     ROE    ROA    EPS  \
MSFT              44.30         68.76   28.90  16.49  14.11   
AAPL              31.89         46.91  151.91  31.18   7.55   
GOOGL             39.55         59.17   32.12  23.16  10.27   
AMZN              13.76         50.05   20.69  10.51   7.17   
META              44.41         82.00   30.16  19.26  23.30   

### Analyze revenue growth trends across companies


In [3]:
# Analyze quarterly revenue trends and R&D investments
quarterly_data = {
    'MSFT': {
        'Q1_2025': {'revenue': 69.632, 'rd': 7.917, 'net_income': 24.108},
        'Q2_2025': {'revenue': 70.066, 'rd': 8.198, 'net_income': 25.824},
        'Q3_2025': {'revenue': 76.441, 'rd': 8.829, 'net_income': 27.233},
        'Q4_2025': {'revenue': 77.673, 'rd': 8.146, 'net_income': 27.747}
    },
    'AAPL': {
        'Q1_2025': {'revenue': 124.300, 'rd': 8.268, 'net_income': 36.330},
        'Q2_2025': {'revenue': 95.359, 'rd': 8.550, 'net_income': 24.780},
        'Q3_2025': {'revenue': 94.036, 'rd': 8.866, 'net_income': 23.434},
        'Q4_2025': {'revenue': 102.466, 'rd': 8.866, 'net_income': 27.466}
    },
    'GOOGL': {
        'Q4_2024': {'revenue': 96.469, 'rd': 13.116, 'net_income': 26.536},
        'Q1_2025': {'revenue': 90.234, 'rd': 13.556, 'net_income': 34.540},
        'Q2_2025': {'revenue': 96.428, 'rd': 13.808, 'net_income': 28.196},
        'Q3_2025': {'revenue': 102.346, 'rd': 15.151, 'net_income': 34.979}
    },
    'AMZN': {
        'Q4_2024': {'revenue': 187.792, 'rd': 23.571, 'net_income': 20.004},
        'Q1_2025': {'revenue': 155.667, 'rd': 22.994, 'net_income': 17.127},
        'Q2_2025': {'revenue': 167.702, 'rd': 27.166, 'net_income': 18.164},
        'Q3_2025': {'revenue': 180.169, 'rd': 28.962, 'net_income': 21.187}
    },
    'META': {
        'Q4_2024': {'revenue': 48.385, 'rd': 12.180, 'net_income': 20.838},
        'Q1_2025': {'revenue': 42.314, 'rd': 12.150, 'net_income': 16.644},
        'Q2_2025': {'revenue': 47.516, 'rd': 12.942, 'net_income': 18.337},
        'Q3_2025': {'revenue': 51.242, 'rd': 15.144, 'net_income': 2.709}
    },
    'NVDA': {
        'Q4_2025': {'revenue': 39.331, 'rd': 3.714, 'net_income': 22.091},
        'Q1_2026': {'revenue': 44.062, 'rd': 3.989, 'net_income': 18.775},
        'Q2_2026': {'revenue': 46.743, 'rd': 4.291, 'net_income': 26.422},
        'Q3_2026': {'revenue': 57.006, 'rd': 4.705, 'net_income': 31.910}
    }
}

# Create dataframes for analysis
revenue_data = []
rd_data = []
profit_data = []

for company in companies:
    for quarter, values in quarterly_data[company].items():
        revenue_data.append({
            'Company': company,
            'Quarter': quarter,
            'Revenue': values['revenue']
        })
        rd_data.append({
            'Company': company,
            'Quarter': quarter,
            'R&D': values['rd'],
            'R&D_as_pct_revenue': (values['rd'] / values['revenue']) * 100
        })
        profit_data.append({
            'Company': company,
            'Quarter': quarter,
            'Net_Income': values['net_income'],
            'Net_Margin': (values['net_income'] / values['revenue']) * 100
        })

df_revenue = pd.DataFrame(revenue_data)
df_rd = pd.DataFrame(rd_data)
df_profit = pd.DataFrame(profit_data)

# Calculate YoY growth and total R&D spending
print("=" * 80)
print("REVENUE AND R&D ANALYSIS (Last 4 Quarters, $ Billions)")
print("=" * 80)

summary_stats = []
for company in companies:
    company_rev = df_revenue[df_revenue['Company'] == company]['Revenue'].values
    company_rd = df_rd[df_rd['Company'] == company]['R&D'].values
    
    total_revenue = company_rev.sum()
    total_rd = company_rd.sum()
    avg_rd_pct = (total_rd / total_revenue) * 100
    
    # Calculate quarter-over-quarter growth
    if len(company_rev) >= 2:
        qoq_growth = ((company_rev[-1] - company_rev[0]) / company_rev[0]) * 100
    else:
        qoq_growth = 0
    
    summary_stats.append({
        'Company': company_names[company],
        'Total Revenue': total_revenue,
        'Total R&D': total_rd,
        'R&D % of Revenue': avg_rd_pct,
        'Revenue Growth (QoQ)': qoq_growth
    })

df_summary = pd.DataFrame(summary_stats)
print(df_summary.round(2))
print("\n")

# Save summary
df_summary.to_csv('artifacts/revenue_rd_analysis.csv', index=False)
print("✓ Saved revenue and R&D analysis to artifacts/revenue_rd_analysis.csv")


REVENUE AND R&D ANALYSIS (Last 4 Quarters, $ Billions)
             Company  Total Revenue  Total R&D  R&D % of Revenue  \
0          Microsoft         293.81      33.09             11.26   
1              Apple         416.16      34.55              8.30   
2  Alphabet (Google)         385.48      55.63             14.43   
3             Amazon         691.33     102.69             14.85   
4    Meta (Facebook)         189.46      52.42             27.67   
5             NVIDIA         187.14      16.70              8.92   

   Revenue Growth (QoQ)  
0                 11.55  
1                -17.57  
2                  6.09  
3                 -4.06  
4                  5.90  
5                 44.94  


✓ Saved revenue and R&D analysis to artifacts/revenue_rd_analysis.csv



### Analyze market performance (stock price movements, returns)


In [4]:
# Analyze stock market performance over the past year
# Calculate returns and volatility

stock_performance = {
    'MSFT': {'start_price': 418.58, 'end_price': 472.39, 'year_high': 555.45, 'year_low': 367.24},
    'AAPL': {'start_price': 243.85, 'end_price': 270.48, 'year_high': 277.7, 'year_low': 169.21},
    'GOOGL': {'start_price': 189.43, 'end_price': 315.32, 'year_high': 328.83, 'year_low': 140.53},
    'AMZN': {'start_price': 220.22, 'end_price': 226.49, 'year_high': 258.6, 'year_low': 161.38},
    'META': {'start_price': 599.24, 'end_price': 652.28, 'year_high': 790.8, 'year_low': 479.8},
    'NVDA': {'start_price': 138.31, 'end_price': 189.29, 'year_high': 212.19, 'year_low': 86.62}
}

performance_data = []
for ticker, data in stock_performance.items():
    year_return = ((data['end_price'] - data['start_price']) / data['start_price']) * 100
    max_gain = ((data['year_high'] - data['start_price']) / data['start_price']) * 100
    max_drawdown = ((data['year_low'] - data['start_price']) / data['start_price']) * 100
    
    performance_data.append({
        'Company': company_names[ticker],
        'Start Price': data['start_price'],
        'End Price': data['end_price'],
        'Year Return (%)': year_return,
        'Year High': data['year_high'],
        'Year Low': data['year_low'],
        'Max Gain (%)': max_gain,
        'Max Drawdown (%)': max_drawdown,
        'Volatility Range (%)': ((data['year_high'] - data['year_low']) / data['start_price']) * 100
    })

df_performance = pd.DataFrame(performance_data)

print("=" * 80)
print("STOCK MARKET PERFORMANCE ANALYSIS (Past Year)")
print("=" * 80)
print(df_performance.round(2))
print("\n")

# Rank companies by performance
print("PERFORMANCE RANKINGS:")
print("-" * 80)
df_ranked = df_performance.sort_values('Year Return (%)', ascending=False)
for idx, row in df_ranked.iterrows():
    print(f"{row['Company']:20s} | Return: {row['Year Return (%)']:6.2f}% | "
          f"High: ${row['Year High']:7.2f} | Low: ${row['Year Low']:7.2f}")
print("\n")

# Save performance data
df_performance.to_csv('artifacts/stock_performance_analysis.csv', index=False)
print("✓ Saved stock performance analysis to artifacts/stock_performance_analysis.csv")


STOCK MARKET PERFORMANCE ANALYSIS (Past Year)
             Company  Start Price  End Price  Year Return (%)  Year High  \
0          Microsoft       418.58     472.39            12.86     555.45   
1              Apple       243.85     270.48            10.92     277.70   
2  Alphabet (Google)       189.43     315.32            66.46     328.83   
3             Amazon       220.22     226.49             2.85     258.60   
4    Meta (Facebook)       599.24     652.28             8.85     790.80   
5             NVIDIA       138.31     189.29            36.86     212.19   

   Year Low  Max Gain (%)  Max Drawdown (%)  Volatility Range (%)  
0    367.24         32.70            -12.27                 44.96  
1    169.21         13.88            -30.61                 44.49  
2    140.53         73.59            -25.81                 99.40  
3    161.38         17.43            -26.72                 44.15  
4    479.80         31.97            -19.93                 51.90  
5     86.62  

### Create comparative visualizations (revenue growth, R&D spending, profitability, stock performance)


In [5]:
# Create comprehensive visualizations
fig, axes = plt.subplots(2, 2, figsize=(16, 12))
fig.suptitle('Big Tech AI Companies - Comparative Analysis', fontsize=16, fontweight='bold')

# 1. Market Cap Comparison
ax1 = axes[0, 0]
market_caps = df_metrics['Market Cap (B)'].sort_values(ascending=True)
colors = ['#0078D4', '#A4A4A4', '#4285F4', '#FF9900', '#0668E1', '#76B900']
market_caps.plot(kind='barh', ax=ax1, color=colors)
ax1.set_xlabel('Market Capitalization ($ Billions)', fontsize=11)
ax1.set_title('Market Capitalization Comparison', fontsize=12, fontweight='bold')
ax1.grid(axis='x', alpha=0.3)
for i, v in enumerate(market_caps.values):
    ax1.text(v + 50, i, f'${v:,.0f}B', va='center', fontsize=10)

# 2. Revenue Growth vs Earnings Growth
ax2 = axes[0, 1]
scatter_data = df_metrics[['Revenue Growth', 'Earnings Growth']].copy()
for idx, company in enumerate(scatter_data.index):
    ax2.scatter(scatter_data.loc[company, 'Revenue Growth'], 
               scatter_data.loc[company, 'Earnings Growth'],
               s=300, alpha=0.6, color=colors[idx], label=company)
    ax2.annotate(company, 
                (scatter_data.loc[company, 'Revenue Growth'], 
                 scatter_data.loc[company, 'Earnings Growth']),
                fontsize=9, ha='center', va='bottom')
ax2.axhline(y=0, color='red', linestyle='--', alpha=0.3)
ax2.axvline(x=0, color='red', linestyle='--', alpha=0.3)
ax2.set_xlabel('Revenue Growth (%)', fontsize=11)
ax2.set_ylabel('Earnings Growth (%)', fontsize=11)
ax2.set_title('Revenue Growth vs Earnings Growth', fontsize=12, fontweight='bold')
ax2.grid(alpha=0.3)

# 3. Profitability Margins Comparison
ax3 = axes[1, 0]
margin_data = df_metrics[['Gross Margin', 'Operating Margin', 'Net Margin']]
x = np.arange(len(margin_data.index))
width = 0.25
ax3.bar(x - width, margin_data['Gross Margin'], width, label='Gross Margin', alpha=0.8)
ax3.bar(x, margin_data['Operating Margin'], width, label='Operating Margin', alpha=0.8)
ax3.bar(x + width, margin_data['Net Margin'], width, label='Net Margin', alpha=0.8)
ax3.set_xlabel('Company', fontsize=11)
ax3.set_ylabel('Margin (%)', fontsize=11)
ax3.set_title('Profitability Margins Comparison', fontsize=12, fontweight='bold')
ax3.set_xticks(x)
ax3.set_xticklabels(margin_data.index, rotation=45, ha='right')
ax3.legend(loc='upper right')
ax3.grid(axis='y', alpha=0.3)

# 4. Stock Performance (Year Return)
ax4 = axes[1, 1]
returns = df_performance.set_index('Company')['Year Return (%)'].sort_values(ascending=True)
bar_colors = ['green' if x > 0 else 'red' for x in returns.values]
returns.plot(kind='barh', ax=ax4, color=bar_colors, alpha=0.7)
ax4.set_xlabel('Year-to-Date Return (%)', fontsize=11)
ax4.set_title('Stock Performance (Past Year)', fontsize=12, fontweight='bold')
ax4.axvline(x=0, color='black', linestyle='-', linewidth=0.8)
ax4.grid(axis='x', alpha=0.3)
for i, v in enumerate(returns.values):
    ax4.text(v + 1, i, f'{v:.1f}%', va='center', fontsize=10)

plt.tight_layout()
plt.savefig('artifacts/comparative_analysis_dashboard.png', dpi=300, bbox_inches='tight')
print("✓ Saved comparative analysis dashboard to artifacts/comparative_analysis_dashboard.png")
plt.close()


✓ Saved comparative analysis dashboard to artifacts/comparative_analysis_dashboard.png



### Generate summary statistics and key insights


In [6]:
# Create additional visualizations for R&D and AI investments
fig, axes = plt.subplots(2, 2, figsize=(16, 12))
fig.suptitle('Big Tech AI Companies - R&D Investment & Efficiency Analysis', fontsize=16, fontweight='bold')

# 1. Total R&D Spending (Last 4 Quarters)
ax1 = axes[0, 0]
rd_spending = df_summary.set_index('Company')['Total R&D'].sort_values(ascending=True)
rd_spending.plot(kind='barh', ax=ax1, color='#FF6B6B', alpha=0.7)
ax1.set_xlabel('Total R&D Spending ($ Billions)', fontsize=11)
ax1.set_title('Total R&D Investment (Last 4 Quarters)', fontsize=12, fontweight='bold')
ax1.grid(axis='x', alpha=0.3)
for i, v in enumerate(rd_spending.values):
    ax1.text(v + 1, i, f'${v:.1f}B', va='center', fontsize=10)

# 2. R&D as % of Revenue
ax2 = axes[0, 1]
rd_pct = df_summary.set_index('Company')['R&D % of Revenue'].sort_values(ascending=False)
colors_rd = ['#FF6B6B' if x > 15 else '#4ECDC4' if x > 10 else '#95E1D3' for x in rd_pct.values]
rd_pct.plot(kind='bar', ax=ax2, color=colors_rd, alpha=0.7)
ax2.set_ylabel('R&D as % of Revenue', fontsize=11)
ax2.set_title('R&D Intensity (R&D % of Revenue)', fontsize=12, fontweight='bold')
ax2.set_xticklabels(rd_pct.index, rotation=45, ha='right')
ax2.grid(axis='y', alpha=0.3)
ax2.axhline(y=15, color='red', linestyle='--', alpha=0.3, label='High Investment (>15%)')
ax2.legend()
for i, v in enumerate(rd_pct.values):
    ax2.text(i, v + 0.5, f'{v:.1f}%', ha='center', fontsize=10)

# 3. Return on Equity (ROE) vs Return on Assets (ROA)
ax3 = axes[1, 0]
roe_roa_data = df_metrics[['ROE', 'ROA']].sort_values('ROE', ascending=True)
x = np.arange(len(roe_roa_data.index))
width = 0.35
ax3.barh(x - width/2, roe_roa_data['ROE'], width, label='ROE', alpha=0.8, color='#6C5CE7')
ax3.barh(x + width/2, roe_roa_data['ROA'], width, label='ROA', alpha=0.8, color='#A29BFE')
ax3.set_xlabel('Return (%)', fontsize=11)
ax3.set_title('Return on Equity (ROE) vs Return on Assets (ROA)', fontsize=12, fontweight='bold')
ax3.set_yticks(x)
ax3.set_yticklabels(roe_roa_data.index)
ax3.legend()
ax3.grid(axis='x', alpha=0.3)

# 4. P/E Ratio vs EPS
ax4 = axes[1, 1]
for idx, company in enumerate(df_metrics.index):
    pe = df_metrics.loc[company, 'P/E Ratio']
    eps = df_metrics.loc[company, 'EPS']
    ax4.scatter(pe, eps, s=400, alpha=0.6, color=colors[idx], label=company)
    ax4.annotate(company, (pe, eps), fontsize=9, ha='center', va='bottom')
ax4.set_xlabel('P/E Ratio', fontsize=11)
ax4.set_ylabel('Earnings Per Share ($)', fontsize=11)
ax4.set_title('Valuation: P/E Ratio vs EPS', fontsize=12, fontweight='bold')
ax4.grid(alpha=0.3)
ax4.legend(loc='best', fontsize=8)

plt.tight_layout()
plt.savefig('artifacts/rd_investment_efficiency_analysis.png', dpi=300, bbox_inches='tight')
print("✓ Saved R&D investment & efficiency analysis to artifacts/rd_investment_efficiency_analysis.png")
plt.close()


✓ Saved R&D investment & efficiency analysis to artifacts/rd_investment_efficiency_analysis.png



### Generate summary statistics and key insights


In [7]:
# Generate comprehensive key insights and summary statistics
print("=" * 80)
print("BIG TECH AI COMPANIES - COMPREHENSIVE ANALYSIS SUMMARY")
print("=" * 80)
print("\n")

# Key Insights Section
insights = []

# 1. Market Leadership
print("1. MARKET LEADERSHIP & VALUATION")
print("-" * 80)
top_market_cap = df_metrics['Market Cap (B)'].idxmax()
print(f"   • Largest by Market Cap: {company_names[top_market_cap]} (${df_metrics.loc[top_market_cap, 'Market Cap (B)']:.1f}B)")
print(f"   • Total Combined Market Cap: ${df_metrics['Market Cap (B)'].sum():.1f}B")
print(f"   • Average P/E Ratio: {df_metrics['P/E Ratio'].mean():.1f}x")
print(f"   • Highest P/E: {company_names[df_metrics['P/E Ratio'].idxmax()]} ({df_metrics['P/E Ratio'].max():.1f}x)")
print(f"   • Lowest P/E: {company_names[df_metrics['P/E Ratio'].idxmin()]} ({df_metrics['P/E Ratio'].min():.1f}x)")
print("\n")

# 2. Growth Performance
print("2. GROWTH PERFORMANCE")
print("-" * 80)
fastest_revenue = df_metrics['Revenue Growth'].idxmax()
fastest_earnings = df_metrics['Earnings Growth'].idxmax()
print(f"   • Fastest Revenue Growth: {company_names[fastest_revenue]} ({df_metrics.loc[fastest_revenue, 'Revenue Growth']:.2f}%)")
print(f"   • Fastest Earnings Growth: {company_names[fastest_earnings]} ({df_metrics.loc[fastest_earnings, 'Earnings Growth']:.2f}%)")
print(f"   • Average Revenue Growth: {df_metrics['Revenue Growth'].mean():.2f}%")
print(f"   • Companies with >10% Revenue Growth: {(df_metrics['Revenue Growth'] > 10).sum()}")
print("\n")

# 3. Profitability Analysis
print("3. PROFITABILITY & EFFICIENCY")
print("-" * 80)
highest_margin = df_metrics['Net Margin'].idxmax()
highest_roe = df_metrics['ROE'].idxmax()
print(f"   • Highest Net Margin: {company_names[highest_margin]} ({df_metrics.loc[highest_margin, 'Net Margin']:.1f}%)")
print(f"   • Average Net Margin: {df_metrics['Net Margin'].mean():.1f}%")
print(f"   • Highest ROE: {company_names[highest_roe]} ({df_metrics.loc[highest_roe, 'ROE']:.1f}%)")
print(f"   • Average ROE: {df_metrics['ROE'].mean():.1f}%")
print(f"   • Highest Operating Margin: {company_names[df_metrics['Operating Margin'].idxmax()]} ({df_metrics['Operating Margin'].max():.1f}%)")
print("\n")

# 4. R&D Investment
print("4. R&D & AI INVESTMENT")
print("-" * 80)
highest_rd_total = df_summary.set_index('Company')['Total R&D'].idxmax()
highest_rd_pct = df_summary.set_index('Company')['R&D % of Revenue'].idxmax()
print(f"   • Highest Total R&D: {highest_rd_total} (${df_summary.set_index('Company').loc[highest_rd_total, 'Total R&D']:.1f}B)")
print(f"   • Total Industry R&D: ${df_summary['Total R&D'].sum():.1f}B (last 4 quarters)")
print(f"   • Highest R&D Intensity: {highest_rd_pct} ({df_summary.set_index('Company').loc[highest_rd_pct, 'R&D % of Revenue']:.1f}% of revenue)")
print(f"   • Average R&D Intensity: {df_summary['R&D % of Revenue'].mean():.1f}%")
print("\n")

# 5. Stock Performance
print("5. STOCK MARKET PERFORMANCE (Past Year)")
print("-" * 80)
best_performer = df_performance.set_index('Company')['Year Return (%)'].idxmax()
most_volatile = df_performance.set_index('Company')['Volatility Range (%)'].idxmax()
print(f"   • Best Performer: {best_performer} (+{df_performance.set_index('Company').loc[best_performer, 'Year Return (%)']:.1f}%)")
print(f"   • Average Return: {df_performance['Year Return (%)'].mean():.1f}%")
print(f"   • Positive Returns: {(df_performance['Year Return (%)'] > 0).sum()}/6 companies")
print(f"   • Most Volatile: {most_volatile} ({df_performance.set_index('Company').loc[most_volatile, 'Volatility Range (%)']:.1f}% range)")
print("\n")

# 6. Risk Assessment
print("6. RISK & VOLATILITY ASSESSMENT")
print("-" * 80)
for idx, row in df_performance.iterrows():
    risk_level = "HIGH" if row['Volatility Range (%)'] > 70 else "MODERATE" if row['Volatility Range (%)'] > 50 else "LOW"
    print(f"   • {row['Company']:20s}: {risk_level:8s} (Max Drawdown: {row['Max Drawdown (%)']:6.1f}%)")
print("\n")


BIG TECH AI COMPANIES - COMPREHENSIVE ANALYSIS SUMMARY


1. MARKET LEADERSHIP & VALUATION
--------------------------------------------------------------------------------
   • Largest by Market Cap: NVIDIA ($4603.5B)
   • Total Combined Market Cap: $20007.8B
   • Average P/E Ratio: 34.4x
   • Highest P/E: NVIDIA (46.4x)
   • Lowest P/E: Meta (Facebook) (28.0x)


2. GROWTH PERFORMANCE
--------------------------------------------------------------------------------
   • Fastest Revenue Growth: NVIDIA (13.27%)
   • Fastest Earnings Growth: NVIDIA (14.55%)
   • Average Revenue Growth: 5.39%
   • Companies with >10% Revenue Growth: 1


3. PROFITABILITY & EFFICIENCY
--------------------------------------------------------------------------------
   • Highest Net Margin: NVIDIA (53.0%)
   • Average Net Margin: 31.6%
   • Highest ROE: Apple (151.9%)
   • Average ROE: 57.9%
   • Highest Operating Margin: NVIDIA (62.3%)


4. R&D & AI INVESTMENT
---------------------------------------------------

### Create a comprehensive comparison dashboard


In [8]:
# Create final comprehensive executive summary report
report = []

report.append("=" * 80)
report.append("EXECUTIVE SUMMARY: BIG TECH AI COMPANIES COMPARATIVE ANALYSIS")
report.append("=" * 80)
report.append("")
report.append("Analysis Period: January 2025 - January 2026")
report.append("Companies Analyzed: Microsoft, Apple, Alphabet (Google), Amazon, Meta, NVIDIA")
report.append("")

report.append("=" * 80)
report.append("KEY FINDINGS & STRATEGIC INSIGHTS")
report.append("=" * 80)
report.append("")

report.append("1. AI LEADERSHIP & MARKET DOMINANCE")
report.append("-" * 80)
report.append("   NVIDIA emerges as the clear AI infrastructure leader:")
report.append("   • Largest market cap: $4.6 trillion")
report.append("   • Highest revenue growth: 13.3% (far exceeding peers)")
report.append("   • Exceptional profitability: 53% net margin, 62% operating margin")
report.append("   • Stock return: +37% (2nd best performer)")
report.append("")
report.append("   Key Insight: NVIDIA's dominance in AI chips (GPUs) positions it as the")
report.append("   'picks and shovels' provider of the AI revolution, benefiting from all")
report.append("   other companies' AI investments.")
report.append("")

report.append("2. ALPHABET (GOOGLE): THE SURPRISE WINNER")
report.append("-" * 80)
report.append("   Alphabet delivered the best stock performance:")
report.append("   • Stock return: +66.5% (best among all six)")
report.append("   • Strong revenue growth: 3.8%")
report.append("   • Solid profitability: 32% net margin")
report.append("   • Heavy R&D investment: $55.6B (14.4% of revenue)")
report.append("")
report.append("   Key Insight: Market recognizes Google's AI integration (Gemini, Search AI)")
report.append("   and cloud growth potential. However, high volatility (99% range) indicates")
report.append("   significant uncertainty and risk.")
report.append("")

report.append("3. R&D INVESTMENT PATTERNS")
report.append("-" * 80)
report.append("   Total industry R&D spending: $295.1 billion (last 4 quarters)")
report.append("")
report.append("   Investment Leaders:")
report.append("   • Amazon: $102.7B total (14.9% of revenue) - AWS AI infrastructure")
report.append("   • Alphabet: $55.6B total (14.4% of revenue) - AI models & cloud")
report.append("   • Meta: $52.4B total (27.7% of revenue!) - Metaverse & AI")
report.append("")
report.append("   Key Insight: Meta's extraordinary 27.7% R&D intensity reflects aggressive")
report.append("   AI and metaverse investments, but resulted in -18% earnings growth,")
report.append("   raising questions about ROI and capital efficiency.")
report.append("")

report.append("4. PROFITABILITY & EFFICIENCY ANALYSIS")
report.append("-" * 80)
report.append("   Profitability Champions:")
report.append("   • NVIDIA: 53% net margin (exceptional for hardware)")
report.append("   • Microsoft: 36% net margin (software/cloud efficiency)")
report.append("   • Alphabet: 32% net margin (advertising + cloud)")
report.append("")
report.append("   Efficiency Leaders (ROE):")
report.append("   • Apple: 152% ROE (capital-light model, massive buybacks)")
report.append("   • NVIDIA: 83% ROE (high-margin AI chips)")
report.append("   • Alphabet: 32% ROE")
report.append("")
report.append("   Concern: Amazon's 11% net margin remains lowest, despite massive scale.")
report.append("   Retail operations continue to pressure overall profitability.")
report.append("")

report.append("5. STOCK PERFORMANCE & RISK PROFILE")
report.append("-" * 80)
report.append("   All six companies delivered positive returns (average: +23.1%)")
report.append("")
report.append("   Performance Ranking:")
report.append("   1. Alphabet: +66.5% (HIGH risk - 99% volatility)")
report.append("   2. NVIDIA: +36.9% (HIGH risk - 91% volatility)")
report.append("   3. Microsoft: +12.9% (LOW risk - 45% volatility)")
report.append("   4. Apple: +10.9% (LOW risk - 44% volatility)")
report.append("   5. Meta: +8.9% (MODERATE risk - 52% volatility)")
report.append("   6. Amazon: +2.9% (LOW risk - 44% volatility)")
report.append("")
report.append("   Key Insight: Higher AI exposure correlates with higher returns BUT also")
report.append("   significantly higher volatility. Conservative investors may prefer")
report.append("   Microsoft or Apple for AI exposure with lower risk.")
report.append("")

report.append("=" * 80)
report.append("INVESTMENT RECOMMENDATIONS BY INVESTOR PROFILE")
report.append("=" * 80)
report.append("")

report.append("GROWTH INVESTORS (High Risk Tolerance):")
report.append("   1. NVIDIA - Pure AI infrastructure play, exceptional growth & margins")
report.append("   2. Alphabet - Strong AI integration, cloud growth, attractive valuation (P/E: 31)")
report.append("   Rationale: Highest growth potential, leading AI capabilities")
report.append("")

report.append("BALANCED INVESTORS (Moderate Risk):")
report.append("   1. Microsoft - Steady growth, Azure AI leadership, enterprise dominance")
report.append("   2. Apple - Strong ecosystem, emerging AI features, capital returns")
report.append("   Rationale: Solid fundamentals, lower volatility, consistent performance")
report.append("")

report.append("VALUE/INCOME INVESTORS (Low Risk):")
report.append("   1. Apple - Lowest volatility, strong cash flow, shareholder returns")
report.append("   2. Microsoft - Predictable revenue, dividend growth, enterprise stability")
report.append("   Rationale: Defensive qualities, proven business models, lower drawdowns")
report.append("")

report.append("AVOID/CAUTION:")
report.append("   • Meta - Negative earnings growth (-18%), unclear metaverse ROI")
report.append("   • Amazon - Lowest margins, modest returns despite massive R&D")
report.append("")

report.append("=" * 80)
report.append("STRATEGIC THEMES & FUTURE OUTLOOK")
report.append("=" * 80)
report.append("")

report.append("1. AI Infrastructure Dominance: NVIDIA's GPU monopoly creates pricing power")
report.append("   and exceptional margins. Watch for competition from custom chips (Google TPU,")
report.append("   Amazon Trainium, Microsoft Maia).")
report.append("")

report.append("2. Cloud AI Integration: Microsoft (Azure AI), Alphabet (Vertex AI), and")
report.append("   Amazon (Bedrock) are racing to monetize AI through cloud platforms.")
report.append("   Microsoft currently leads in enterprise AI adoption.")
report.append("")

report.append("3. Consumer AI Applications: Apple's on-device AI and Google's Search AI")
report.append("   integration will drive next phase of growth. Watch for AI-driven")
report.append("   hardware refresh cycles.")
report.append("")

report.append("4. R&D Efficiency Concerns: Meta's 27.7% R&D spending with negative earnings")
report.append("   growth raises questions. Investors should monitor ROI on AI investments")
report.append("   across all companies.")
report.append("")

report.append("5. Regulatory Risks: All six face antitrust scrutiny. AI regulation could")
report.append("   impact competitive dynamics and profitability.")
report.append("")

report.append("=" * 80)
report.append("CONCLUSION")
report.append("=" * 80)
report.append("")
report.append("The Big Tech AI race is creating unprecedented value ($20 trillion combined")
report.append("market cap) but with significant differentiation:")
report.append("")
report.append("• NVIDIA: Clear AI infrastructure winner - highest growth & profitability")
report.append("• Microsoft: Best balanced play - growth + stability + enterprise AI")
report.append("• Alphabet: Highest upside potential but with elevated risk")
report.append("• Apple: Defensive quality play with emerging AI catalysts")
report.append("• Meta: High risk turnaround story - execution concerns remain")
report.append("• Amazon: Solid but unexciting - margins remain a challenge")
report.append("")
report.append("All six companies are investing heavily in AI ($295B combined R&D), but")
report.append("monetization strategies and execution vary significantly. Investors should")
report.append("align choices with risk tolerance and investment horizon.")
report.append("")
report.append("=" * 80)
report.append("END OF REPORT")
report.append("=" * 80)

# Save report
report_text = "\n".join(report)
with open('artifacts/executive_summary_report.txt', 'w') as f:
    f.write(report_text)

print(report_text)
print("\n✓ Saved executive summary report to artifacts/executive_summary_report.txt")


EXECUTIVE SUMMARY: BIG TECH AI COMPANIES COMPARATIVE ANALYSIS

Analysis Period: January 2025 - January 2026
Companies Analyzed: Microsoft, Apple, Alphabet (Google), Amazon, Meta, NVIDIA

KEY FINDINGS & STRATEGIC INSIGHTS

1. AI LEADERSHIP & MARKET DOMINANCE
--------------------------------------------------------------------------------
   NVIDIA emerges as the clear AI infrastructure leader:
   • Largest market cap: $4.6 trillion
   • Highest revenue growth: 13.3% (far exceeding peers)
   • Exceptional profitability: 53% net margin, 62% operating margin
   • Stock return: +37% (2nd best performer)

   Key Insight: NVIDIA's dominance in AI chips (GPUs) positions it as the
   'picks and shovels' provider of the AI revolution, benefiting from all
   other companies' AI investments.

2. ALPHABET (GOOGLE): THE SURPRISE WINNER
--------------------------------------------------------------------------------
   Alphabet delivered the best stock performance:
   • Stock return: +66.5% (best amo

## Final Plan Status

```
1. [x] Identify major Big Tech AI companies and retrieve their current stock price snapshots
2. [x] Retrieve historical stock price data for the past year for each company
3. [x] Retrieve financial statements (income statements, balance sheets, cash flow statements) for each company
4. [x] Retrieve financial metrics snapshots for key performance indicators
5. [x] Load and explore all retrieved data
6. [x] Analyze revenue growth trends across companies
7. [x] Analyze R&D investments and profitability metrics
8. [x] Analyze market performance (stock price movements, returns)
9. [x] Create comparative visualizations (revenue growth, R&D spending, profitability, stock performance)
10. [x] Generate summary statistics and key insights
11. [x] Create a comprehensive comparison dashboard
12. [x] Provide final summary with actionable insights and recommendations
```

**All 12 steps completed!**


---

## Final Answer

# Big Tech AI Companies - Comprehensive Comparative Analysis

## Executive Summary

I've completed a comprehensive comparative analysis of six major Big Tech AI companies (Microsoft, Apple, Alphabet/Google, Amazon, Meta, and NVIDIA) covering the period from January 2025 to January 2026. The analysis examined revenue growth, R&D investments, profitability, and market performance.

## Key Findings

### 1. **Market Leadership & Valuation**
- **NVIDIA** leads with the largest market cap at **$4.6 trillion**
- Combined market capitalization: **$20 trillion**
- All six companies delivered positive returns, averaging **+23.1%**

### 2. **Growth Performance**
- **NVIDIA** dominates with **13.3% revenue growth** and **14.6% earnings growth**
- **Alphabet** delivered the best stock performance: **+66.5% return**
- Only NVIDIA achieved >10% revenue growth, highlighting its AI infrastructure dominance

### 3. **R&D Investment & AI Spending**
- Total industry R&D: **$295.1 billion** (last 4 quarters)
- **Amazon** leads in absolute spending: **$102.7B**
- **Meta** has the highest R&D intensity: **27.7% of revenue** (concerning given -18% earnings growth)
- Average R&D intensity: **14.2% of revenue**

### 4. **Profitability Analysis**
- **NVIDIA** leads in margins: **53% net margin, 62% operating margin**
- **Apple** has exceptional ROE: **152%** (due to capital-light model and buybacks)
- **Amazon** lags with only **11% net margin** despite massive scale

### 5. **Stock Performance & Risk**
**Performance Ranking:**
1. Alphabet: +66.5% (HIGH volatility - 99% range)
2. NVIDIA: +36.9% (HIGH volatility - 91% range)
3. Microsoft: +12.9% (LOW volatility - 45% range)
4. Apple: +10.9% (LOW volatility - 44% range)
5. Meta: +8.9% (MODERATE volatility - 52% range)
6. Amazon: +2.9% (LOW volatility - 44% range)

## Investment Recommendations by Profile

### **Growth Investors (High Risk Tolerance)**
1. **NVIDIA** - Pure AI infrastructure play with exceptional growth & margins
2. **Alphabet** - Strong AI integration, attractive P/E ratio (31x)

### **Balanced Investors (Moderate Risk)**
1. **Microsoft** - Azure AI leadership, enterprise dominance, steady growth
2. **Apple** - Strong ecosystem, emerging AI features, capital returns

### **Value/Income Investors (Low Risk)**
1. **Apple** - Lowest volatility, strong cash flow, shareholder returns
2. **Microsoft** - Predictable revenue, dividend growth, stability

### **Caution/Avoid**
- **Meta** - Negative earnings growth (-18%), unclear metaverse ROI
- **Amazon** - Lowest margins, modest returns despite massive R&D

## Strategic Insights

1. **AI Infrastructure Dominance**: NVIDIA's GPU monopoly creates exceptional pricing power and margins, positioning it as the "picks and shovels" provider of the AI revolution

2. **Cloud AI Race**: Microsoft (Azure AI), Alphabet (Vertex AI), and Amazon (Bedrock) are competing to monetize AI through cloud platforms

3. **R&D Efficiency Concerns**: Meta's 27.7% R&D spending with negative earnings growth raises serious ROI questions

4. **Risk-Return Tradeoff**: Higher AI exposure correlates with higher returns but significantly higher volatility

5. **Regulatory Risks**: All six face antitrust scrutiny and potential AI regulation

## Deliverables Created

All analysis artifacts have been saved to the `artifacts/` folder:
- ✅ `financial_metrics_summary.csv` - Key financial metrics comparison
- ✅ `revenue_rd_analysis.csv` - Revenue and R&D investment analysis
- ✅ `stock_performance_analysis.csv` - Market performance metrics
- ✅ `comparative_analysis_dashboard.png` - 4-panel visualization dashboard
- ✅ `rd_investment_efficiency_analysis.png` - R&D and efficiency analysis charts
- ✅ `executive_summary_report.txt` - Comprehensive written report

## Conclusion

NVIDIA emerges as the clear AI infrastructure winner with the highest growth and profitability. Microsoft offers the best balanced play combining growth with stability. Alphabet presents the highest upside potential but with elevated risk. Apple provides defensive quality with emerging AI catalysts. Meta and Amazon face execution challenges despite massive investments.

The $295B combined R&D investment demonstrates the industry's commitment to AI, but monetization strategies and execution vary significantly. Investors should align their choices with their risk tolerance and investment horizon.
